pkg/pipelineascode/template.go

```diff
@@ -0,0 +1,181 @@
+	}
+
+	prg, err := env.Program(checked, cel.EvalOptions(cel.OptOptimize))
+	if err != nil {
+		return nil, fmt.Errorf("expression %#v failed to create a Program: %w", expr, err)
+	}
+
+	out, _, err := prg.Eval(data)
+	if err != nil {
+		return nil, fmt.Errorf("expression %#v failed to evaluate: %w", expr, err)
+	}
+
+	return out, nil
+}
+
+// celFilter filters a query with cel, it sets two variables for the user to use
+// in the cel filter. body and pac.
+// body is the payload of the event, since it's already casted we un-marshall/and
+// marshall it as a map[string]interface{}
+func celFilter(query string, body any, pacParams map[string]string) (bool, error) {
+	nbody, err := json.Marshal(body)
+	if err != nil {
+		return false, err
+	}
+	var jsonMap map[string]interface{}
+	err = json.Unmarshal(nbody, &jsonMap)
+	if err != nil {
+		return false, err
+	}
+
+	mapStrDyn := decls.NewMapType(decls.String, decls.Dyn)
+	celDec, _ := cel.NewEnv(
+		cel.Declarations(
+			decls.NewVar("body", mapStrDyn),
+			decls.NewVar("pac", mapStrDyn),
+		))
+	val, err := celEvaluate(query, celDec, map[string]any{
+		"body": jsonMap,
+		"pac":  pacParams,
+	})
+	if err != nil {
+		return false, err
+	}
+
+	//nolint: gocritic
+	switch val.(type) {
+	case types.Bool:
+		if val.Value() == true {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// processParams will process the parameters as set in the repo.Spec CR.
+// value can come from a string or from a secretKeyRef or from a string value
+// if both is set we pick the value and issue a warning in the user namespace
+// we let the user specify a cel filter. If false then we skip the parameters.
+// if multiple params name has a filter we pick up the first one that has
+// matched true.
+func (p *PacRun) processParams(ctx context.Context, repo *v1alpha1.Repository, maptemplate map[string]string) error {
+	mapFilters := map[string]string{}
+	if repo.Spec.Params == nil {
+		return nil
+	}
+	for index, value := range *repo.Spec.Params {
+		// if the name is empty we skip it
+		if value.Name == "" {
+			p.eventEmitter.EmitMessage(repo, zap.ErrorLevel,
+				"ParamsFilterSkipped", fmt.Sprintf("no name has been set in params[%d] of repo %s", index, repo.GetName()))
+			continue
+		}
+		if value.Filter != "" {
+			// if we already have a filter that has matched we skip it
+			if _, ok := mapFilters[value.Name]; ok {
+				p.eventEmitter.EmitMessage(repo, zap.WarnLevel,
+					"ParamsFilterSkipped", fmt.Sprintf("skipping params name %s, filter has already been matched previously", value.Name))
+				continue
+			}
+
+			// if the cel filter condition is false we skip it
+			cond, err := celFilter(value.Filter, p.event.Event, maptemplate)
+			if err != nil {
+				return err
+			}
+			if !cond {
+				p.eventEmitter.EmitMessage(repo, zap.InfoLevel,
+					"ParamsFilterSkipped", fmt.Sprintf("skipping params name %s, filter condition is false", value.Name))
+				continue
+			}
+			mapFilters[value.Name] = value.Value
```

Comment:
What if value is `" "` ?

Reply:
ah yeah, it would be just case... like the same way if you didn't have anything.. it could be used with filters to set the variable to empty if it match a filter.. maybe i should document that

Reply:
I would prefer to document it so that there won't be confusions